## Формула расчёта выборки для A/B тестирования.

### Формула позволяет рассчитать количество клиентов в 2-х выборках из общего числа клиентов сегмента для A/B теста так, чтобы:
- полученные результаты учитывали минимально значимое различие в конверсии,
- полученные результаты учитывали необходимый уровень значимости.

**В формуле участвуют следующие элементы:**

*Выборка (sample)* - это минимально необходимое количество клиентов только в одной группе для проведения теста с заданными критериями.

*Нулевая и альтернативная гипотезы* - так как численное значение, которое суммирует интересующие нас различия называется эффектом, для того, чтобы определить, есть он или нет выдвигают 2 гипотезы - нулевую и альтернативную.  Обычно нулевая гипотеза гласит, что эффекта нет (клиенты, на которых воздействовали реагируют также, как и те, на которых воздействия не было). Альтернативная наоборот, признаёт наличие определённого эффекта, хотя о полном доказательстве альтернативной гипотезы по результатам выборки речи не идёт. 

*Доверительная вероятность (р-оценка)* - это вероятность наблюдения в исследовании такого же или более сильного эффекта при условии справедливости нулевой гипотезы. Обычно выражается как пропорция. Значение доверительной оценки тесно связано с уровнем значимости.

*Уровень значимости (confidence level)* - это пороговое значение для р-оценки, ниже которого нулевая гипотеза должна быть отвергнута и сделано заключение о том, что имеются доказательства эффекта. Обычно уровень значимости устанавливается на значении 5%. Уровень значимости, несмотря на прямую связь с р-оценкой выражается в процентах: 5% уровень значимости эквивалентен р=0.05).

*Мощность (power)* - это вероятность того, что нулевая гипотеза будет адекватно отвергнута, иными словами тогда, когда действительно существуют доказательства реальных различий или взаимосвязей. Ее можно рассматривать как "100 процентов минус вероятность пропуска истинного эффекта", поэтому чем выше мощность, тем меньше вероятность пропуска истинного эффекта. Мощность обычно фиксируется на уровне 80%, 90% или 95%. В данной формуле мощность была зафиксирована на минимально необходимом уровне 85%, более высокое значение параметра привело бы к дополнительному разрастанию выборки.

*Ожидаемая конверсия* - это процент конверсии, который мы ожидаем от первой и от второй выборки. Она устанавливается вручную на основе исторических данных о реакции данного сегмента. Чем выше ожидаемая конверсия, тем больше выборка нам потребуется. 

*Минимально значимое различие в конверсии* - иногда в сравнении результатов двух выборок в A/B тесте мы можем получить статистически значимое, но реально - несущественное различие, для этого в формуле участвует параметр минимально значимого ращличия. Благодаря ему мы можем установить, что, например, минимально значимым различием в конверсии двух выборок будет 5%, и в связке с ожидаемой конверсией по двум выборкам - узнать, какое нам необходимо для этого количество человек.

In [4]:
import math

In [3]:
def ab_sample_count(confLevel, min_rate, pA, pB):
    if confLevel == 1:
        Zalpha = 2.326
        Zbeta = 2.576
    else:
        Zalpha = 1.645
        Zbeta = 1.96
    
    sample = ((Zalpha+Zbeta)**2)*( (pA*(1-pA)) + (pB*(1-pB)) / (min_rate**2))
    corrected_sample = round(sample+(sample/2.2)+0.5)
    
    print('Уровень значимости: {}%.'.format(100 - confLevel))
    print('Минимально значимое различие в конверсии: {}%.'.format(int(min_rate*100)))
    print('Ожидаемая конверсия: группа A - {}%, группа B - {}%.'.format(int(pA*100), int(pB*100)))
    print('Выборка для одной группы: {}.'.format(corrected_sample))
    print('Полная выборка для A/B тестирования: {}.'.format(corrected_sample*2))

### Пример использования формулы:
**Задача:** необходимо рассчитать 2 выборки с целью провести на них A/B тестирование и узнать, значимо ли реально различие между результатами двух выборок при условии, что, отсекать гипотезу относительно ожидаемого эффекта будем по значению 5%. Ожидаемым эффектом и нулевой гипотезой таким образом будем считать конверсию в 12% в выборках А и B по отдельности, а минимально значимое различие конверсий установим на 5%.

Задаём параметры для расчёта данной выборки:
- **confLevel - это уровень значимости:** допустимые для формулы значения 1 и 5, по ним рассчитывается статистический критерий Z. При необходимости можно расширить диапазон рассчитываемых значений для критерия Z, но как правило уровень значимости оставляют равным либо 1 либо 5.
- **min_rate - это минимально значимое различие в конверсиях** выборок А и B, это 5% и задаётся долей как 0.05
- **pA - это ожидаемая конверсия группы A,** она равна 12% и задаётся долей как 0.12
- **pB - это ожидаемая конверсия группы B,** она тоже равна 12% и задаётся долей как 0.12

Параметры для расчёта передаются в словарь. Сразу после этого запускаем формулу с нашим словарём параметров, чтобы увидеть результат:

In [5]:
sample_params = {
    'confLevel': 5,
    'min_rate': 0.05,
    'pA': 0.12,
    'pB': 0.12,
}

ab_sample_count(**sample_params)

Уровень значимости: 95%.
Минимально значимое различие в конверсии: 5%.
Ожидаемая конверсия: группа A - 12%, группа B - 12%.
Выборка для одной группы: 801.
Полная выборка для A/B тестирования: 1602.


**Результат:** необходимое число людей в выборке для A/B теста составляет 801 человека, а полная выборка A/B тестирования из всей совокупности клиентов составит 1602 человека.